### PR TITLE
[Routing] force stringable object to string in route parameter

### LINKED ISSUE #26992: [Routing] extra parameters objects should be casted to string
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | any

Real case scenario: I want to pass an extra parameter to a route (see [docs](http://symfony.com/doc/current/routing.html#generating-urls-with-query-strings)), and the value of such parameter is an object. The typical example is a `Uuid` object (e.g. Ramsey\Uuid).
Problem is that PHP is not smart enough and it does not cast objects to string. See [related bug on php.net](https://bugs.php.net/bug.php?id=66966).
Of couse, such feature could be easily obtained in a typical plain PHP context, since it's very easy to cast your object to a string.
In a typical Symfony context, instead, you can't easily cast strings, unless doing ugly things.

Example of not working code:
```twig
{# generated URL has no query string #}
<a href="{{ path('my_route', {foo: object.uuid}) }}">my link</a>
```

Example of working code, with ugly workaround:
```twig
{# generated URL has expected query string, e.g. ?foo=7ecff38e-4145-11e8-80b9-0242ac140004 #}
<a href="{{ path('my_route', {foo: object.uuid.__toString}) }}">my link</a>
```
My proposal is: inside `doGenerate` method of `UrlGenerator` class, Symfony should cast parameters to string before passing them to `http_build_query`

### PR DESCRIPTION
**Pull request by garak**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #26992
| License       | MIT
| Doc PR        | none

This is a simple proposal to fix related issue.

### CODE REVIEW

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 1]
```diff
@@ -172,6 +172,9 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $variables = array_flip($variables);
         $mergedParams = array_replace($defaults, $this->context->getParameters(), $parameters);
 
+        // force string for an object. See https://bugs.php.net/bug.php?id=66966
+        $parameters = array_map(static function ($param) { return \is_object($param) ? (string) $param : $param; }, $parameters);
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: I wonder if you should perform this operation on the `$extra` array instead, which holds only those parameters that will be funneled into `http_build_query()`.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 2]
```diff
@@ -294,6 +294,9 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             return $a == $b ? 0 : 1;
         });
 
+        // force string for an object. See https://bugs.php.net/bug.php?id=66966
+        $extra = array_map(static function ($param) { return \is_object($param) ? (string) $param : $param; }, $extra);
```
Comment (Nyholm) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: Maybe we should check if the `__toString()` method exists? Ie, if we can cast an object to a string.

Comment (garak) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: I wasn't sure about taking the burden of such control.
What would be the benefit? And what the library should do if a user passes a non-stringable object?

Comment (Nyholm) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: The current implementation will cause an Error: 

https://3v4l.org/m6ovP

I think we should throw an exception. The benefit would be that we are in the control of the exception message and may help the user to find and fix the problem.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 3]
```diff
@@ -186,9 +186,9 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
                 // variable is not important by default
                 $important = $token[5] ?? false;
 
-                if (!$optional || $important || !\array_key_exists($varName, $defaults) || (null !== $mergedParams[$varName] && (string) $mergedParams[$varName] !== (string) $defaults[$varName])) {
+                if (!$optional || $important || !\array_key_exists($varName, $defaults) || (null !== $mergedParams[$varName] && (string)$mergedParams[$varName] !== (string)$defaults[$varName])) {
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: CS changes should be reverted.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 4]
```diff
@@ -273,27 +273,39 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
...
         });
 
+        // force string for an object. See https://bugs.php.net/bug.php?id=66966
+        foreach ($extra as $paramName => $param) {
+            if (!\is_object($param)) {
+                continue;
+            }
+            if (!\is_callable([$param, '__toString'])) {
+                $message = 'Cannot convert parameter "{parameter}" to string.';
+                throw new InvalidParameterException(strtr($message, '{parameter}', $paramName));
```
Comment (Nyholm) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: Could we have the message saying something like: 

`$x = get_debug_type($param)`
"Error when generating an URL for "$name": Cannot convert parameter "$paramName" to string, "$x"  given. 

Or is this overkill?

(obviously with nice style, correct grammar and `sprintf`)

**Review by Nyholm — CHANGES_REQUESTED** [review]
Nice!

Sorry if Fabbot suggested a bunch of CS changes. :/

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 5]
```diff
@@ -294,6 +294,18 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
...
         });
 
+        // force string for an object. See https://bugs.php.net/bug.php?id=66966
+        foreach ($extra as $paramName => $param) {
+            if (!\is_object($param)) {
+                continue;
+            }
+            if (!\is_callable([$param, '__toString'])) {
+                $message = 'Error when generating an URL for "{route}". Cannot convert object of type {type} to string for parameter "{parameter}".';
+                throw new InvalidParameterException(strtr($message, ['{route}' => $name, '{type}' => get_debug_type($param), '{parameter}' => $paramName]));
```
Comment (Nyholm) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: Why not use `sprintf()`, we use that everywhere else, dont we?

```suggestion
                throw new InvalidParameterException(\sprintf('Error when generating an URL for "%s". Cannot convert object of type "%s" to string for parameter "%s".', $name, get_debug_type($param), $paramName));
```

File: src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php [hunk 6]
```diff
@@ -856,6 +856,23 @@ public function testFragmentsCanBeDefinedAsDefaults()
         $this->assertEquals('/app.php/testing#fragment', $url);
     }
 
+    public function testNonStringableObjectParameterCannotBeConvertedToString()
+    {
+        $this->expectException(InvalidParameterException::class);
+        $object = new \stdClass();
+        $routes = $this->getRoutes('test', new Route('/testing'));
```
Comment (Nyholm) [inline src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php]:
```suggestion
        $object = new \stdClass();
        $routes = $this->getRoutes('test', new Route('/testing'));
        $this->expectException(InvalidParameterException::class);
```

I had a discussion with @wouterj about this the other day. See https://github.com/symfony/symfony/pull/40284#discussion_r582939157

**Review by Nyholm — COMMENTED** [review]
Thank you. 
Im happy with this after two small changes.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 5 above)
Comment (garak) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: Indeed, the same file is using `sprintf` 2 times and `strtr` 2 times

File: src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php (hunk 6 above)
Comment (garak) [reply src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php]: I just tried to stay consistent with other uses of the same method in the file.
But I agree that expectation near statement is better.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 5 above)
Comment (Nyholm) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: I quick (and maybe flawed) regex search showed that we are only using `strtr` when throwing an exception twice in this repo... both in this file =)

Let's use `sprintf`.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 7]
```diff
@@ -294,6 +294,17 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             return $a == $b ? 0 : 1;
         });
 
+        // force string for an object. See https://bugs.php.net/bug.php?id=66966
+        foreach ($extra as $paramName => $param) {
+            if (!\is_object($param)) {
+                continue;
+            }
+            if (!\is_callable([$param, '__toString'])) {
```
Comment (Nyholm) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php:302]: I think a future PR can deal with if you pass an array or callable. 
Or we may think that is not needed.

**Review by Nyholm — APPROVED** [review]
Thank you. 

Im happy with this.

**Comment by Nyholm** [comment]
Oh, Fabbot saw that you missed something when adding the sprintf. :)

**Review by derrabus — APPROVED** [review]

**Comment by fabpot** [comment]
I think I agree with @linaori's comments in the related issue, so mostly :-1: 
In any case, that would be for 5.x.

**Comment by fabpot** [comment]
I forgot to mention that we tried to remove such string casts in the past from the framework. IIRC, @Tobion was quite vocal about this.

**Comment by Tobion** [comment]
👎 https://github.com/symfony/symfony/issues/26992#issuecomment-791094799

**Comment by fabpot** [comment]
Closing as 2 core team members voted against this change.

**Pull request closed without merging**